backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,195 @@
+    const sessionId = record.sessionId?.S;
+    if (!sessionId) {
+      return errorResult({
+        errorMessage: "Session is malformed",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+
+    return successResult(sessionId);
+  }
+
+  async readSessionDetails(
+    subjectIdentifier: string,
+  ): Promise<Result<SessionDetails | null>> {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: "ASYNC_AUTH_SESSION_CREATED" },
+        ":currentTimeInSeconds": {
+          N: currentTimeInSeconds.toString(),
+        },
+      },
+      ExpressionAttributeNames: { "#state": "state" },
+      ProjectionExpression: "sessionId, redirectUri, #state",
+      Limit: 1,
+      ScanIndexForward: false,
+    };
+
+    const queryResponse = await this.query(input);
+    if (queryResponse.isError || queryResponse.value === null) {
+      return queryResponse;
+    }
+
+    const record: DynamoDbRecord = queryResponse.value;
+    const sessionId = record.sessionId?.S;
+    const state = record.state?.S;
+    if (!sessionId || !state) {
+      return errorResult({
+        errorMessage: "Session is malformed",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+    const redirectUri: string | undefined = record.redirectUri?.S;
+    const sessionDetails: SessionDetails = {
+      sessionId,
+      state,
+      ...(redirectUri && { redirectUri: record.redirectUri?.S }),
+    };
+
+    return successResult(sessionDetails);
+  }
+
+  async createSession(
+    attributes: CreateSessionAttributes,
+    sessionId: string,
+  ): Promise<Result<string>> {
+    const {
+      client_id,
+      govuk_signin_journey_id,
+      issuer,
+      redirect_uri,
+      sessionDurationInSeconds,
+      state,
+      sub,
+    } = attributes;
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const timeToLive = currentTimeInSeconds + sessionDurationInSeconds;
+
+    const input: PutItemCommandInput = {
+      TableName: this.tableName,
+      Item: {
+        clientId: { S: client_id },
+        govukSigninJourneyId: { S: govuk_signin_journey_id },
+        createdAt: { N: Date.now().toString() },
+        issuer: { S: issuer },
+        sessionId: { S: sessionId },
+        sessionState: { S: "ASYNC_AUTH_SESSION_CREATED" },
+        state: { S: state },
+        subjectIdentifier: { S: sub },
+        timeToLive: { N: timeToLive.toString() },
+      },
+      ConditionExpression: "attribute_not_exists(sessionId)",
+    };
+
+    if (redirect_uri) {
+      input.Item!.redirectUri = { S: redirect_uri };
```

Comment:
it feels very strange you have to have the `!` 🤔

Reply:
I know :(

Reply:
Is it better now? @sandymay-dd